### PR TITLE
feat: adding live-balances to bPool

### PR DESCRIPTION
**Pull request by wei3erHase**
Closes BAL-25

### CODE REVIEW

**Comment by wei3erHase** [comment]
Comments on #15:
> Add documentation on why making this change: [automatic gulping](https://github.com/defi-wonderland/balancer-v1-amm/pull/15#pullrequestreview-2058715596).
> [Could remove `gulp(token)` method](https://github.com/defi-wonderland/balancer-v1-amm/pull/15#discussion_r1602069836)

File: test/unit/BPool.t.sol [hunk 1]
```diff
@@ -54,6 +54,10 @@ abstract contract BasePoolTest is Test, BConst, Utils, BMath {
     vm.mockCall(_token, abi.encodeWithSelector(IERC20(_token).transferFrom.selector), abi.encode(true));
   }
 
+  function _mockPoolBalance(address _token, uint256 _balance) internal {
+    vm.mockCall(_token, abi.encodeWithSelector(IERC20(_token).balanceOf.selector, address(bPool)), abi.encode(_balance));
```
Comment (0xAustrian) [inline test/unit/BPool.t.sol:58]: The selector can be simplified to simply be `IERC20.balanceOf.selector`

**Review by 0xAustrian — APPROVED** [review]
Small optional comment, looks good!

File: test/unit/BPool.t.sol [hunk 2]
```diff
@@ -651,6 +652,7 @@ contract BPool_Unit_Bind is BasePoolTest {
   function _setValues(Bind_FuzzScenario memory _fuzz) internal {
     // Create mocks
     _mockTransferFrom(_fuzz.token);
+    _mockPoolBalance(_fuzz.token, 0);
```
Comment (wei3erHase) [inline test/unit/BPool.t.sol:655]: needed to mock the call even tho it returns 0
we shall remove it on #54

File: test/unit/BPool.t.sol [hunk 3]
```diff
@@ -1055,7 +1051,7 @@ contract BPool_Unit_Unbind is BasePoolTest {
     assertEq(bPool.call__records(_fuzz.previousTokens[_fuzz.tokenIndex]).index, 0);
     assertEq(bPool.call__records(_fuzz.previousTokens[_fuzz.tokenIndex]).bound, false);
     assertEq(bPool.call__records(_fuzz.previousTokens[_fuzz.tokenIndex]).denorm, 0);
-    assertEq(bPool.call__records(_fuzz.previousTokens[_fuzz.tokenIndex]).balance, 0);
+    // assertEq(bPool.call__records(_fuzz.previousTokens[_fuzz.tokenIndex]).balance, 0);
```
Comment (wei3erHase) [inline test/unit/BPool.t.sol:1054]: removing commented line

File: test/unit/BPool.t.sol [hunk 4]
```diff
@@ -2435,6 +2316,7 @@ contract BPool_Unit_JoinswapExternAmountIn is BasePoolTest {
   }
 
   function test_Revert_MaxInRatio(JoinswapExternAmountIn_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    vm.skip(true); // TODO: FIX! this test is failing
```
Comment (0xAustrian) [inline test/unit/BPool.t.sol:2319]: Create a Linear task for this and I will take a look later, but it's weird that fails now.

File: test/unit/BPool.t.sol [hunk 5]
```diff
@@ -3146,6 +2989,7 @@ contract BPool_Unit_ExitswapExternAmountOut is BasePoolTest {
   }
 
   function test_Revert_MaxOutRatio(ExitswapExternAmountOut_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    vm.skip(true); // TODO: FIX! this test is failing
```
Comment (0xAustrian) [inline test/unit/BPool.t.sol:2992]: Same as above.

**Review by 0xAustrian — APPROVED** [review]

File: test/unit/BPool.t.sol (hunk 4 above)
Comment (wei3erHase) [reply test/unit/BPool.t.sol:2319]: Added BAL-75

**Pull request merged**